# Kinetics of Xist-induced gene silencing depend on combinations of epigenetic and genomic features

Published in Genome Research (2019)

PMID: 31175153


### Abstract
To initiate X Chromosome inactivation (XCI), the long non-coding RNA Xist mediates chromosome-wide gene silencing of one X Chromosome in female mammals to equalize gene dosage between the sexes. The efficiency of gene silencing is highly variable across genes, with some genes even escaping XCI in somatic cells. A geneâ€™s susceptibility to Xist-mediated silencing appears to be determined by a complex interplay of epigenetic and genomic features; however, the underlying rules remain poorly understood. We have quantified chromosome-wide gene silencing kinetics at the level of the nascent transcriptome using allele-specific Precision nuclear Run-On sequencing (PRO-seq). We have developed a Random Forest machine learning model that can predict the measured silencing dynamics based on a large set of epigenetic and genomic features and tested its predictive power experimentally. The genomic distance to the Xist locus, followed by gene density and distance to LINE elements are the prime determinants of the speed of gene silencing. Moreover, we find two distinct gene classes associated with different silencing pathways: a class that requires Xist repeat A for silencing, which is known to activate the SPEN-pathway and a second cluster where genes are pre-marked by Polycomb complexes and tend to rely on the B repeat in Xist for silencing, known to recruit polycomb complexes during XCI. Moreover, a series of features associated with active transcriptional elongation and chromatin 3D structure are enriched at rapidly silenced genes. Our machine learning approach can thus uncover the complex combinatorial rules underlying gene silencing during X inactivation.

### Content

This contains all data sources and scripts that were used to perform all analysis my thesis, based on the analysis for the paper (the paper has a seperate repositoty on the marsicoLab page).

The folder data contains all required processed data (big raw files are not included), as well as data of modelling results.
The folder scripts contains all scripts needed to reproduce the obtained results, including pre-processing steps, modelling and interpretation steps.
The folder plots contains all plots generated for the thesis.



